src/pmdas/amdgpu/amdgpu.c

```diff
@@ -0,0 +1,668 @@
+    if ((name = strdup(gpuname)) == NULL) {
+      pmNoMem("gcard instname", strlen(gpuname), PM_RECOV_ERR);
+      while (--i)
+        free(pcp_amdgpuinfo.indom->it_set[i].i_name);
+      free(pcp_amdgpuinfo.devs);
+      free(pcp_amdgpuinfo.indom->it_set);
+      free(pcp_amdgpuinfo.info);
+      return -ENOMEM;
+    }
+    pcp_amdgpuinfo.indom->it_set[i].i_name = name;
+
+    info->fd = -1;
+    getAMDDevice(dev, &info->amd_device, &info->fd);
+    /* Get static values */
+    if (DRMDeviceGetGPUInfo(info->amd_device, &info->gpu_info)) {
+	info->failed[MEMORY_CLUSTER][AMDGPU_MEMORY_CLOCK_MAX] = 1;
+	info->failed[GPU_CLUSTER][AMDGPU_GPU_CLOCK_MAX] = 1;
+    }
+  }
+
+  pcp_amdgpuinfo.indom->it_numinst = device_count;
+  return 0;
+}
+
+static int refresh(pcp_amdgpuinfo_t *amdgpuinfo, uint32_t to_refresh)
+{
+  int i;
+
+  if (!drm_initialized) {
+      setup_gcard_indom();
+      drm_initialized = 1;
+  }
+
+  for (i = 0; i < amdgpuinfo->numcards && i < amdgpuinfo->maxcards; i++) {
+      drmMemory_t memory = {0};
+      amdgpu_info_t *info = &amdgpuinfo->info[i];
+      void *param = NULL;
+      int ret = DRM_SUCCESS;
+
+      switch (to_refresh) {
+	case AMDGPU_NAME_REFRESHER:
+	  param = info->name;
+	  break;
+	case AMDGPU_GPU_INFO_REFRESHER:
+	  param = &info->gpu_info;
+	  break;
+	case AMDGPU_GPU_CLOCK_REFRESHER:
+	  param = &info->gpu_clock;
+	  break;
+	case AMDGPU_GPU_TEMPERATURE_REFRESHER:
+	  param = &info->temperature;
+	  break;
+	case AMDGPU_GPU_LOAD_REFRESHER:
+	  param = &info->load;
+	  break;
+	case AMDGPU_GPU_AVG_PWR_REFRESHER:
+	  param = &info->avg_pwr;
+	  break;
+	case AMDGPU_MEMORY_INFO_REFRESHER:
+	  param = &memory;
+	  break;
+	case AMDGPU_MEMORY_CLOCK_REFRESHER:
+	  param = &info->mem_clock;
+	  break;
+      }
+
+      ret = amd_refresher[to_refresh].refresher(info->amd_device, param);
+
+      if (ret != DRM_SUCCESS) {
+	  /* Mark all metrics that depend on the same refresher as failed */
+	  for (int j = 0; j < CLUSTER_COUNT; j++)
+	    for (int k = 0; k < MAX_ITEM_COUNT; k++) {
+		if (&amd_refresher[to_refresh] == refresher_list[j][k])
+		  info->failed[j][k] = 1;
+	    }
+
+	  continue;
+      }
+
+      if (param == &memory) {
+	  info->memory = memory; /* struct copy */
+	  info->memaccum += memory.used;
+	  pmNotifyErr(LOG_ERR, "Refreshed memory (%lx)", memory.used);
+      }
+  }
+
+  return 0;
+}
+
+static int amdgpu_instance(pmInDom indom, int inst, char *name,
+                           pmInResult **result, pmdaExt *pmda) {
+  return pmdaInstance(indom, inst, name, result, pmda);
+}
+
+/*
+ * Wrapper for pmdaFetch which refresh the set of values once per fetch
+ * PDU.  The fetchCallback is then called once per-metric/instance pair
+ * to perform the actual filling of the pmResult (via each pmAtomValue).
+ */
+static int amdgpu_fetch(int numpmid, pmID pmidlist[], pmResult **resp,
+                        pmdaExt *pmda)
+{
+  uint32_t i = 0;
+
+  pmNotifyErr(LOG_ERR, "amdgpu_fetch");
```

Comment:
Leftover from debugging?  Alternatively, call it conditionally under a pmDebugOptions guard (off by default)